tower-router/src/lib.rs

```diff
@@ -0,0 +1,398 @@
+//! # use futures_util::never::Never;
+//! # use tower_router::ServiceRouter;
+//! type StdError = Box<dyn std::error::Error + Send + Sync + 'static>;
+//! struct MyService(u8);
+//!
+//! impl Service<String> for MyService {
+//!     type Response = ();
+//!     type Error = StdError;
+//!     type Future = Ready<Result<(), Self::Error>>;
+//!
+//!     fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+//!         Poll::Ready(Ok(()))
+//!     }
+//!
+//!     fn call(&mut self, req: String) -> Self::Future {
+//!         println!("{}: {}", self.0, req);
+//!         ready(Ok(()))
+//!     }
+//! }
+//!
+//! #[tokio::main]
+//! async fn main() {
+//!     let mut s = ServiceRouter::new(
+//!         vec![MyService(0), MyService(1)],
+//!         // one service handles strings with uppercase first letters. the other handles the rest.
+//!         |r: &String| if r.chars().next().unwrap().is_uppercase() { 0 } else { 1 },
+//!     );
+//!
+//!     let reqs = vec!["A", "b", "C", "d"];
+//!     let reqs: Vec<String> = reqs.into_iter().map(String::from).collect();
+//!     for r in reqs {
+//!         poll_fn(|cx| s.poll_ready(cx)).await.unwrap();
+//!         s.call(r).await;
+//!     }
+//! }
+//! ```
+#![deny(missing_docs)]
+#![warn(unreachable_pub)]
+#![warn(missing_copy_implementations)]
+#![warn(trivial_casts)]
+#![warn(trivial_numeric_casts)]
+#![warn(unused_extern_crates)]
+#![warn(rust_2018_idioms)]
+#![warn(missing_debug_implementations)]
+#![allow(clippy::type_complexity)]
+
+use futures_util::stream::Stream;
+use std::sync::Arc;
+use std::{
+    collections::VecDeque,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::sync::{oneshot, Mutex};
+use tower_service::Service;
+
+type StdError = Box<dyn std::error::Error + Send + Sync + 'static>;
+
+/// This is how callers of [`ServiceRouter`] tell it which `Service` a `Req` corresponds to.
+pub trait Router<Req> {
+    /// Return an index into the iterator of `Service` passed to [`ServiceRouter::new`].
+    fn route(&mut self, r: &Req) -> usize;
+}
+
+impl<F, Req> Router<Req> for F
+where
+    F: Fn(&Req) -> usize,
+{
+    fn route(&mut self, r: &Req) -> usize {
+        self(r)
+    }
+}
+
+/// `ServiceRouter` manages a list of `Service`s.
+///
+/// It accepts new requests, then:
+/// 1. Determines, via the provided [`Router`], which `Service` the request coresponds to.
+/// Then it returns a future that:
+/// 2. Waits for that `Service` to be ready (using [`ServiceGroup`]).
+/// 3. Calls the `Service` with the request, and waits for the response.
+/// 4. Resolves to the response of the call.
+///
+/// Unlike `tower-buffer`, `ServiceRouter` does not require `S` to `impl Clone`, but like
+/// `tower-buffer` it uses a [`oneshot`] to manage requests.
+#[derive(Debug)]
+pub struct ServiceRouter<S, Req, T: 'static, F> {
+    router: F,
+    inner: Arc<Mutex<ServiceGroup<OneshotService<S>, (oneshot::Sender<T>, Req)>>>,
```

Comment:
Why the arc mutex?

Reply:
It'll go away. I chatted to @akshayknarayan offline.